### PR TITLE
Customize MethodHandle for map-to-map cast during function specialization

### PR DESCRIPTION
**Pull request by shixuan-fan**
Addresses #9066

### CODE REVIEW

**Comment by shixuan-fan** [comment]
Not ready for review yet

**Comment by martint** [comment]
I'd rephrase the commit title as "Customize methodhandle for map to map cast during function specialization"

And in the description, explain that this improves performance.

**Comment by shixuan-fan** [comment]
This is now ready for review @wenleix @haozhun  :) 

Benchmark result:
Before 2nd commit:
Benchmark                        Mode  Cnt     Score    Error  Units
BenchmarkMapToMapCast.benchmark  avgt   30  2496.135 ± 91.159  ns/op

After 2nd commit:
Benchmark                        Mode  Cnt     Score    Error  Units
BenchmarkMapToMapCast.benchmark  avgt   30  2208.322 ± 68.906  ns/op

Will do further benchmarking if any significant change is made during review process. Thanks!

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MapToMapCast.java [hunk 1]
```diff
@@ -13,70 +13,181 @@
...
+            else {
+                throw new IllegalArgumentException("Unknown java type " + castResultJavaType + " from type " + toType);
+            }
+            target = compose(nullCheck, target);
+        }
+
+        MethodHandle writer = nativeValueWriter(toType);
+        writer = permuteArguments(writer, methodType(void.class, writer.type().parameterArray()[1], writer.type().parameterArray()[0]), 1, 0);
+
+        // The composed method handle could be invoked by: (fromMap, index, connectorSession, blockBuilder)
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MapToMapCast.java]: Put this in the javadoc of the function instead.

In addition, you want to expand it a bit. For example, `The signature of the returned MethodHandle is (Block fromMap, int position, ConnectorSession session, BlockBuilder mapBlockBuilder)void`

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MapToMapCast.java [hunk 2]
```diff
@@ -13,70 +13,181 @@
...
+            cast = MethodHandles.dropArguments(cast, 0, ConnectorSession.class);
+        }
+        cast = permuteArguments(cast, methodType(cast.type().returnType(), cast.type().parameterArray()[1], cast.type().parameterArray()[0]), 1, 0);
+        MethodHandle target = compose(cast, getter);
+
+        // We only need to check null if the method handle is for key and the cast result is nullable.
+        Class<?> castResultJavaType = target.type().returnType();
+        if (isKey && !castResultJavaType.isPrimitive()) {
+            MethodHandle nullCheck;
+            if (castResultJavaType == Long.class) {
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MapToMapCast.java]: Put this `if` in a separate method. Probably a private `nullChecker` method in this class. Add javadoc to that method to explain what it does, and what the signature is.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MapToMapCast.java [hunk 3]
```diff
@@ -13,70 +13,181 @@
...
+        MethodHandle target = MethodHandles.insertArguments(METHOD_HANDLE, 0, keyProcessor, valueProcessor, toMapType);
+        return new ScalarFunctionImplementation(true, ImmutableList.of(valueTypeArgumentProperty(RETURN_NULL_ON_NULL)), target, true);
+    }
+
+    private MethodHandle getProcessMethodHandle(FunctionRegistry functionRegistry, Type fromType, Type toType, boolean isKey)
+    {
+        MethodHandle getter = nativeValueGetter(fromType);
+
+        // Adapt cast that takes ([ConnectorSession,] ?) to one that takes (?, ConnectorSession), where ? is the return type of getter.
+        MethodHandle cast = functionRegistry.getScalarFunctionImplementation(internalOperator(CAST, toType, ImmutableList.of(fromType))).getMethodHandle();
```
[the remark this answers is not in the document]
Comment (shixuan-fan) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/MapToMapCast.java]: Addressed using `getCoercion` as offline discussion.

**Comment by shixuan-fan** [comment]
Before 2nd commit:
Benchmark                        Mode  Cnt     Score    Error  Units
BenchmarkMapToMapCast.benchmark  avgt   30  2211.045 ± 38.092  ns/op

After 2nd commit:
Benchmark                        Mode  Cnt     Score    Error  Units
BenchmarkMapToMapCast.benchmark  avgt   30  1930.737 ± 33.572  ns/op

New benchmark shows ~12.7% improvement.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MapToMapCast.java [hunk 4]
```diff
@@ -13,70 +13,208 @@
...
+        writer = permuteArguments(writer, methodType(void.class, writer.type().parameterArray()[1], writer.type().parameterArray()[0]), 1, 0);
+
+        return compose(writer, target.asType(methodType(unwrap(target.type().returnType()), target.type().parameterArray())));
+    }
+
+    /**
+     * Returns a null checker MethodHandle that only returns the value when it is not null.
+     * The signature of the returned MethodHandle could be one of the following depending on javaType:
+     *   - (Long value)long
+     *   - (Double value)double
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MapToMapCast.java]: http://blog.joda.org/2012/11/javadoc-coding-standards.html
```
  /**
   * First paragraph.
   * <p><ul>
   * <li>the first item
   * <li>the second item
   * <li>the third item
   * </ul><p>
   * Second paragraph.
   */

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MapToMapCast.java [hunk 5]
```diff
@@ -13,70 +13,208 @@
...
+        // Adapt cast that takes ([ConnectorSession,] ?) to one that takes (?, ConnectorSession), where ? is the return type of getter.
+        ScalarFunctionImplementation castImplementation = functionRegistry.getScalarFunctionImplementation(functionRegistry.getCoercion(fromType, toType));
+        MethodHandle cast = castImplementation.getMethodHandle();
+        if (cast.type().parameterArray()[0] != ConnectorSession.class) {
+            cast = MethodHandles.dropArguments(cast, 0, ConnectorSession.class);
+        }
+        cast = permuteArguments(cast, methodType(cast.type().returnType(), cast.type().parameterArray()[1], cast.type().parameterArray()[0]), 1, 0);
+        MethodHandle target = compose(cast, getter);
+
+        // We only need to check null if the method handle is for key and the cast result is nullable.
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MapToMapCast.java]: nit: 
```
If the key cast function is nullable, check the result is not null.
```

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/BenchmarkMapToMapCast.java [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+    {
+        private Page page;
+        private PageProcessor pageProcessor;
+
+        @Setup
+        public void setup()
+        {
+            Signature signature = new Signature("$operator$CAST", FunctionKind.SCALAR, mapType(BIGINT, DOUBLE).getTypeSignature(), mapType(DOUBLE, BIGINT).getTypeSignature());
+
+            List<RowExpression> projections = ImmutableList.of(
```
Comment (wenleix) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/BenchmarkMapToMapCast.java]: Ideally, we want to test different keyTypes and valueTypes, this is because if we only benchmark one fixed type, JVM might do some optimistic optimization for if-statement.

You can find examples in `BenchmarkJsonToMapCast`.

Let's merge this for now and left this as future work. This could boost the performance improvement :).

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MapToMapCast.java [hunk 7]
```diff
@@ -13,70 +13,208 @@
...
+                        TypeSignatureParameter.of(toValueType.getTypeSignature())));
+
+        MethodHandle keyProcessor = buildProcessor(functionRegistry, fromKeyType, toKeyType, true);
+        MethodHandle valueProcessor = buildProcessor(functionRegistry, fromValueType, toValueType, false);
+        MethodHandle target = MethodHandles.insertArguments(METHOD_HANDLE, 0, keyProcessor, valueProcessor, toMapType);
+        return new ScalarFunctionImplementation(true, ImmutableList.of(valueTypeArgumentProperty(RETURN_NULL_ON_NULL)), target, true);
+    }
+
+    /**
+     * The signature of the returned MethodHandle is (Block fromMap, int position, ConnectorSession session, BlockBuilder mapBlockBuilder)void.
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MapToMapCast.java]: Add the following comment
```
The processor will get the value from fromMap, cast it and write to the toBlock
```
